### PR TITLE
randomize exponential backoffs

### PR DESCRIPTION
**Pull request by akosyakov**
## Description

This PR adds jitteriness to the exponential backoff on reconnection to the server and port exposure. It should help to prevent synchronous spiking of port exposure on the webapp deployment.

## Related Issue(s)

Fixes #

## How to test

- Start a workspace and observe it logs `kubectl logs ws-.... -f > ws.log` from dev workspace for this PR
- Scale down servers `kubectl scale --replicas=0 deployment server` from dev workspace for this PR
- Try to trigger the port exposure with `curl lama.sh | sh -s -- -p 3333` from a test worksapce
- Observe in logs that a workspace is trying to reconnect to the server and expose port with random exponential backoff
Screenshot 2022-09-29 at 14 35 23

- Scale up and check that workspace is reconnected `kubectl scale --replicas=1 deployment server` from dev workspace for this PR
- Do it again if you wan to be sure

## Release Notes

```release-note
NONE
```

## Werft options:

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`

### CODE REVIEW

File: components/gitpod-protocol/go/reconnecting-ws.go [hunk 1]
```diff
@@ -214,25 +224,20 @@ func (rc *ReconnectingWebsocket) connect(ctx context.Context) *WebsocketConnecti
...
+			Error("failed to connect, trying again...")
 		select {
 		case <-rc.closedCh:
 			return nil
 		case <-time.After(delay):
-			delay = time.Duration(float64(delay) * rc.reconnectionDelayGrowFactor)
-			if delay > rc.maxReconnectionDelay {
-				delay = rc.maxReconnectionDelay
-			}
+			attempt++
```
Comment (easyCZ) [inline components/gitpod-protocol/go/reconnecting-ws.go:240]: What if we used a package for this? It's way cleaner and offers the same configuration we need.
https://github.com/cenkalti/backoff/blob/v4.1.3/exponential.go#L54

Comment (akosyakov) [reply components/gitpod-protocol/go/reconnecting-ws.go:240]: Can we do it in another PR? It takes time to test and I would like to focus on other things. I'm pretty sure this code works.

Comment (easyCZ) [reply components/gitpod-protocol/go/reconnecting-ws.go:240]: Sure, up to you. Just feels the current version has higher risk of going wrong.

Comment (akosyakov) [reply components/gitpod-protocol/go/reconnecting-ws.go:240]: I'm not sure what you mean. I've implemented Full Jitter from https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/ which seems to easies and perform better. 

The algorithm from https://github.com/cenkalti/backoff/blob/v4.1.3/exponential.go#L54 seems to be equal jitter based on randomness around moving interval. From the article above it performs worse since it is less random.

I can do a change tomorrow morning.

Comment (easyCZ) [reply components/gitpod-protocol/go/reconnecting-ws.go:240]: Sorry, it was less about performance and more about using a tested package over our implementation which has higher chance of going wrong - we've not run it in anger in production.

File: components/supervisor/hot-deploy.sh [hunk 2]
```diff
@@ -18,9 +18,9 @@ dev_image="$(tar xfO "$bldfn" ./imgnames.txt | head -n1)"
...
 
 
-cf_patch=$(kubectl get cm server-ide-config -o=json | jq '.data."config.json"' |jq -r)
+cf_patch=$(kubectl get cm ide-config -o=json | jq '.data."config.json"' |jq -r)
 cf_patch=$(echo "$cf_patch" |jq ".supervisorImage = \"$dev_image\"")
 cf_patch=$(echo "$cf_patch" |jq tostring)
 cf_patch="{\"data\": {\"config.json\": $cf_patch}}"
 
-kubectl patch cm server-ide-config --type=merge -p "$cf_patch"
+kubectl patch cm ide-config --type=merge -p "$cf_patch"
```
Comment (mustard-mh) [inline components/supervisor/hot-deploy.sh:26]: it would be good if we can restart `ide-service` to get this configmap change immediately, see [doc](https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#mounted-configmaps-are-updated-automatically)
i.e. `kubectl rollout restart deployment ide-service`

File: components/supervisor/pkg/ports/exposed-ports.go [hunk 3]
```diff
@@ -153,20 +155,24 @@ func (g *GitpodExposedPorts) doExpose(req *exposePortRequest) {
...
 	attempt := 0
+	base := float64(g.minExposeDelay)
 	for {
 		_, err = g.C.OpenPort(req.ctx, g.WorkspaceID, req.port)
-		if err == nil || req.ctx.Err() != nil || attempt == 5 {
+		if err == nil || req.ctx.Err() != nil || attempt == int(g.maxExposeAttempts) {
 			return
 		}
-		log.WithError(err).WithField("port", req.port).Warnf("cannot expose port, trying again in %d seconds...", uint32(delay.Seconds()))
+		delay := time.Duration(base + rand.Float64()*(base*math.Pow(g.exposeDelayGrowFactor, float64(attempt))))
```
Comment (mustard-mh) [inline components/supervisor/pkg/ports/exposed-ports.go:165]: I search code, it looks like we don't init rand seed in supervisor, so maybe it a pseudo random

Comment (akosyakov) [reply components/supervisor/pkg/ports/exposed-ports.go:165]: I will switch tomorrow to the library as suggested.

File: components/gitpod-protocol/go/reconnecting-ws.go [hunk 4]
```diff
@@ -181,7 +175,13 @@ func (rc *ReconnectingWebsocket) Dial(ctx context.Context) error {
...
 
 func (rc *ReconnectingWebsocket) connect(ctx context.Context) *WebsocketConnection {
-	delay := rc.minReconnectionDelay
+	exp := backoff.NewExponentialBackOff()
+	exp.InitialInterval = 2 * time.Second
+	exp.RandomizationFactor = 0.5
+	exp.MaxInterval = 30 * time.Second
+	exp.Multiplier = 1.5
+	exp.MaxElapsedTime = 0
+	exp.Reset()
```
Comment (jenting) [inline components/gitpod-protocol/go/reconnecting-ws.go:184]: I do not quite understand why we need to call `Reset()` function. Could you please explain a bit more? Thank you.

Comment (akosyakov) [reply components/gitpod-protocol/go/reconnecting-ws.go:184]: because docs says so: https://github.com/cenkalti/backoff/blob/6b0e4ad0cd65431b217393bff47d1dff727b264b/exponential.go#L108

otherwise startTime is not initialized

**Comment by akosyakov** [comment]
@easyCZ @mustard-mh I've switched to the proposed package and retested it. Could you have a look again please? 🙏

File: components/gitpod-protocol/go/reconnecting-ws.go [hunk 5]
```diff
@@ -205,7 +205,13 @@ func (rc *ReconnectingWebsocket) connect(ctx context.Context) *WebsocketConnecti
...
 		}
 
-		if err == websocket.ErrBadHandshake {
+		var statusCode int
+		if resp != nil {
+			statusCode = resp.StatusCode
+		}
+
+		// 200 is bad gateway for ws, we should keep trying
+		if err == websocket.ErrBadHandshake && statusCode != 200 {
```
Comment (easyCZ) [inline components/gitpod-protocol/go/reconnecting-ws.go:214]: After we do the nil check, we access the `resp` below in 
```
if rc.badHandshakeCount > rc.badHandshakeMax || (http.StatusMultipleChoices <= resp.StatusCode && resp.StatusCode < http.StatusInternalServerError) {
				_ = rc.closeWithError(&ErrBadHandshake{rc.url, rc.reqHeader, resp})
				return nil
			}
```
The access to `resp` happens irrespective if `resp` was nil. Here, we could potentially get a panic from `resp.StatusCode`

File: components/gitpod-protocol/go/reconnecting-ws.go [hunk 6]
```diff
@@ -214,25 +220,18 @@ func (rc *ReconnectingWebsocket) connect(ctx context.Context) *WebsocketConnecti
...
-		rc.log.WithField("url", rc.url).Info("websocket handshake")
 
+		delay := exp.NextBackOff()
 		rc.log.WithError(err).
 			WithField("url", rc.url).
 			WithField("badHandshakeCount", fmt.Sprintf("%d/%d", rc.badHandshakeCount, rc.badHandshakeMax)).
 			WithField("statusCode", statusCode).
-			Errorf("failed to connect, trying again in %d seconds...", uint32(delay.Seconds()))
+			WithField("delay", delay.String()).
+			Error("failed to connect, trying again...")
```
Comment (easyCZ) [inline components/gitpod-protocol/go/reconnecting-ws.go:230]: Do we ever stop retrying?

File: components/supervisor/pkg/supervisor/supervisor.go [hunk 7]
```diff
@@ -1101,20 +1102,23 @@ func startAPIEndpoint(ctx context.Context, cfg *Config, wg *sync.WaitGroup, serv
...
 	)
+	unaryInterceptors = append(unaryInterceptors, grpcMetrics.UnaryServerInterceptor())
+	streamInterceptors = append(streamInterceptors, grpcMetrics.StreamServerInterceptor())
+
 	opts = append(opts,
-		grpc.StreamInterceptor(grpcMetrics.StreamServerInterceptor()),
-		grpc.UnaryInterceptor(grpcMetrics.UnaryServerInterceptor()),
+		grpc.UnaryInterceptor(grpc_middleware.ChainUnaryServer(unaryInterceptors...)),
+		grpc.StreamInterceptor(grpc_middleware.ChainStreamServer(streamInterceptors...)),
 	)
```
Comment (easyCZ) [inline components/supervisor/pkg/supervisor/supervisor.go:1122]: This seems unrelated to the exponential backoff change, should it be here?

**Review by easyCZ — APPROVED** [review]
LGTM in general, couple of comments and questions for my own understanding.

/hold for Qs

File: components/gitpod-protocol/go/reconnecting-ws.go (hunk 6 above)
Comment (akosyakov) [reply components/gitpod-protocol/go/reconnecting-ws.go:230]: if a workspace times out or we are receiving valid bad handshakes, i.e. our client is not compatible somehow when it does make sense to retry.

**Review by Pothulapati — APPROVED** [review]
:shipit:

**Review by jenting — APPROVED** [review]

**Comment by felladrin** [comment]
Last build failed [[1](https://werft.gitpod-dev.com/job/gitpod-build-ak-random-backoff.14/)], so I'll try building it again.
/werft run with-clean-slate-deployment=true

**Comment by felladrin** [comment]
ℹ️ I'm reviewing it.

**Review by felladrin — APPROVED** [review]
Unfortunately, I wasn't able to go through the tests. The preview environment keeps getting stuck, and I wonder if I'm testing it correctly. I'll have a meeting with Anton tomorrow morning and will check it.
